guides/4)integrating_other_frameworks/Gradio_and_Wandb_Integration.md

```diff
@@ -0,0 +1,382 @@
+
+    ```python
+    #@title Finetune StyleGAN
+    #@markdown alpha controls the strength of the style
+    alpha =  1.0 #@param {type:"slider", min:0, max:1, step:0.1}
+    alpha = 1-alpha
+
+
+    #@markdown Tries to preserve color of original image by limiting family of allowable transformations. Set to false if you want to transfer color from reference image. This also leads to heavier stylization
+    preserve_color = True #@param{type:"boolean"}
+    #@markdown Number of finetuning steps. Different style reference may require different iterations. Try 200~500 iterations.
+    num_iter = 200 #@param {type:"number"}
+    #@markdown Log training on wandb and interval for image logging
+    use_wandb = True #@param {type:"boolean"}
+    log_interval = 50 #@param {type:"number"}
+
+
+    samples = []
+    column_names = ["Referece (y)", "Style Code(w)", "Real Face Image(x)"]
+
+
+    if use_wandb:
+        wandb.init(project="JoJoGAN")
+        config = wandb.config
+        config.num_iter = num_iter
+        config.preserve_color = preserve_color
+        wandb.log(
+        {"Style reference": [wandb.Image(transforms.ToPILImage()(target_im))]},
+        step=0)
+        wandb.log({"Gradio panel": wandb.Html('''
+    <link rel="stylesheet" href="https://gradio.s3-us-west-2.amazonaws.com/2.6.2/static/bundle.css">
+    <div id="JoJoGAN-demo"></div>
+    <script src="https://gradio.s3-us-west-2.amazonaws.com/2.6.2/static/bundle.js"></script>
+    <script>
+    launchGradioFromSpaces("akhaliq/JoJoGAN", "#JoJoGAN-demo")
+    </script>
+    <style>
+    /* work around a weird bug */
+    .gradio_app .gradio_bg[theme=huggingface] .gradio_interface .input_dropdown .dropdown:hover .dropdown_menu {
+        display: block;
+    }
+    </style>
+    ''', inject=False)})
+
+
+    lpips_fn = lpips.LPIPS(net='vgg').to(device)
+
+
+    # reset generator
+    del generator
+    generator = deepcopy(original_generator)
+
+
+    g_optim = optim.Adam(generator.parameters(), lr=2e-3, betas=(0, 0.99))
+
+
+    # Which layers to swap for generating a family of plausible real images -> fake image
+    if preserve_color:
+        id_swap = [7,9,11,15,16,17]
+    else:
+        id_swap = list(range(7, generator.n_latent))
+
+
+    for idx in tqdm(range(num_iter)):
+        if preserve_color:
+            random_alpha = 0
+        else:
+            random_alpha = np.random.uniform(alpha, 1)
+        mean_w = generator.get_latent(torch.randn([latents.size(0), latent_dim]).to(device)).unsqueeze(1).repeat(1, generator.n_latent, 1)
+        in_latent = latents.clone()
+        in_latent[:, id_swap] = alpha*latents[:, id_swap] + (1-alpha)*mean_w[:, id_swap]
+
+
+        img = generator(in_latent, input_is_latent=True)
+        loss = lpips_fn(F.interpolate(img, size=(256,256), mode='area'), F.interpolate(targets, size=(256,256), mode='area')).mean()
+        
+        if use_wandb:
```

Comment:
Why do we have this as an if statement? Let's just use wandb and gradio since that's the purpose of this blog post

Reply:
removed if statement, thanks